autolab/Autolab
# Global annotations

## Description
- Allow instructors to create, edit, and delete global annotations
- Display all non-autograded problems in annotation pane to facilitate adding global annotations
- Show a dash if a problem is released with no score (via annotation) associated with it. This avoids an internal server error when we try to format `nil`.
- Update annotations documentation (re: shared comments, global annotations, positive grading)
- Fix off-by-one line number on `viewFeedback` page
- Improve validation for max score upon problem creation
- Update logic for `problemReleased` to additionally check that the grading deadline has not passed
- Automatically destroy annotations when a problem is destroyed

## Motivation and Context
To provide instructors with a method of awarding points for a problem globally, without tying it to a specific line of code.

## How Has This Been Tested?
Test using an autograded single file assessment (e.g. hello) and a non-autograded tar assessment
- Ensure that all non-autograded problems are shown in the annotation pane
- Ensure that you can no longer create a problem without a max score
- Ensure that annotation line number shown on viewFeedback page is no longer off-by-one
- Check documentation for annotations
- Test that clickable chevron toggles collapse of a problem
- Successfully create, edit, and delete global annotations
  - Global annotations only appear in the annotation pane, and not in the code viewer
  - Global annotations still influence a student's score correctly
  - Global annotations are not clickable
- Test that above features work even after switching files or adding / editing / deleting annotations
- Ensure that students are unable to view the annotations when either (1) grades are not released and/or (2) the grading deadline has not passed
  - Ensure that students can view annotation otherwise
- Ensure that the view feedback page handles global annotations correctly
  - Clickable link loads the student's submission

**Creating a global annotation**
![Screenshot 2022-11-27 at 16 53 52](https://user-images.githubusercontent.com/9074856/204161657-0c7022c7-8248-45ee-8aac-5a4af5d23f11.png)

![Screenshot 2022-11-27 at 16 54 45](https://user-images.githubusercontent.com/9074856/204161697-f6ca577c-bf59-401e-848c-281b8136073a.png)

**Editing a global annotation**
![Screenshot 2022-11-27 at 16 54 08](https://user-images.githubusercontent.com/9074856/204161664-88150177-537f-477c-810b-ccf8e59f601e.png)

![Screenshot 2022-11-27 at 16 54 52](https://user-images.githubusercontent.com/9074856/204161702-72bbeb0c-86b2-484d-bbbb-fba4121be622.png)


**View feedback page with global annotations** - clicking the link loads the first file of the submission
![Screen Shot 2022-11-02 at 17 49 33](https://user-images.githubusercontent.com/9074856/199609163-a524c2f7-c440-4acd-a3f1-4434e1f0342a.png)

**Student unable to view annotations when unreleased**
![Screen Shot 2022-11-02 at 17 53 30](https://user-images.githubusercontent.com/9074856/199609713-250040fd-b042-4d05-a08b-2b2dbe24a089.png)

![Screen Shot 2022-11-02 at 17 54 25](https://user-images.githubusercontent.com/9074856/199609844-dc0c3373-3b5e-40bf-bc29-65f463b02b8d.png)


## Types of changes

## Checklist:

## Other issues / help required
- When annotations are edited, the textarea is initially too small, but they expand correctly once you start typing. I'm not sure if there's a way to make them have the right height from the start.
- The exception to this is when the file containing the annotation is directly loaded and there have been no actions (e.g. switching files, deleting another annotation, editing another annotation...). In this case, the textarea has a slightly too large height. However, the height does not decrease past the original height when deleting lines.
